### ISSUE TITLE
Why does subMonth not work correctly?

### ISSUE THREAD

**Issue by TheShiftExchange** [briannesbitt/Carbon#428]
It is currently 11:50pm local time on 31st August 2015.

Why do I get the incorrect date when doing subMonth functions? Is Carbon not able to handle 30 and 31 months? :(

```
var_dump(\Carbon\Carbon::now());

Carbon {#520 ▼
  +"date": "2015-08-31 14:00:07.000000"
  +"timezone_type": 3
  +"timezone": "UTC"
}


var_dump(\Carbon\Carbon::now()->subMonth(1));

Carbon {#521 ▼
  +"date": "2015-07-31 14:00:07.000000"
  +"timezone_type": 3
  +"timezone": "UTC"
}

var_dump(\Carbon\Carbon::now()->subMonth(2));

Carbon {#522 ▼
  +"date": "2015-07-01 14:00:07.000000"  // <- WRONG - I would have expected 2015-06-30?
  +"timezone_type": 3
  +"timezone": "UTC"
}

var_dump(\Carbon\Carbon::now()->subMonth(3));

Carbon {#523 ▼
  +"date": "2015-05-31 14:00:07.000000"
  +"timezone_type": 3
  +"timezone": "UTC"
}
```

put another way: The output is "August, July, July, May" - but obviously I want "August, July, June, May".

How do I always safely get the previous X month?

**Comment by briannesbitt** [comment]
Checkout `addMonthsNoOverflow`

**Comment by TheShiftExchange** [comment]
But is it really expected behavior?

If you do `->subMonth(2)->format('M')` - then you would always expect the result to be 2 months earlier... anything else feels wrong?

**Comment by dereuromark** [comment]
@TheShiftExchange While I agree with you it is probably difficult to change the default behavior at this point. And devs need to just know about the overflow issue and use the corresponding method for it.

**Comment by TheShiftExchange** [comment]
Yeah - I understand.

I had a look - doesnt seem to be mentioned anywhere in the docs? Should probably be right next to the add/sub methods so people are away.

Thinking aloud - if I did subYear on a leap year on NYE - I would get the same problem?

**Comment by dereuromark** [comment]
Are there test cases for it? If not they should be added with the "expected" behavior shown for all to see what the intention of the devs is here.
But I think it that you are right and it should be corrected in a major version and the current behavior could be a `addMonthsExact()` method or something.

**Comment by TheShiftExchange** [comment]
@briannesbitt - are you open to changing this in a major tag?

I cannot think of an instance when doing `->subMonth(1)` that it would be acceptable to *not* get the month "1" ago? It would seem a trap that most developers would probably fall into (like I did)?

And what makes it worse - is it is very unlikely to be detected during development, and only show up on the 31st of a month (like what occurred to me).

I reckon it would be a silent bug in many applications? Because doing `->subMonth(1)` would imply you explicitly want to know what the previous month was?

If I was doing `->subDays(30)` - then I am saying that I want 30 days ago - But if I am doing `->subMonth(1)` - then I am explicitly asking for the month ago?

Basically it should always be M-1?

Thoughts?

**Comment by briannesbitt** [comment]
Its based on the underlying PHP behaviour for "consistency".

```
$dt = new DateTime('2015-01-31');
$dt->modify("+1 month");

echo $dt->format("Y-m-d H:i:s");   //2015-03-03 00:00:00
```

Not saying its right or wrong, but you can see its not really adding 30 days but ++ the month and then doing overflow since the resulting day doesn't really exist.

**Comment by TheShiftExchange** [comment]
hmmm... I understand what you are saying.

Does Carbon have 'config' options? If so - could we config 'overflow' to be on/off as default?

**Comment by dereuromark** [comment]
@briannesbitt The point of Carbon among other things is to overwrite the inconsistent or wrong behavior of PHP DateTime, right? So it would make sense to be more consistent with the developer's expectations then the often times quite buggy core behavior IMO.

**Comment by lucasmichot** [comment]
See #710

**Comment by panique** [comment]
This is indeed broken usability at it's finest! 
There's probably no developer in the world who expects 1st of January when doing `subMonth()` on January 31st. This is nuts. To be fair, it's PHP's fault, as the underlaying PHP date engine is doing this horribly wrong.

Hey community, maybe it's time to get it on and fix this shit in PHP8. I'm in :)

**Comment by kylekatarnls** [comment]
Pull requests can be sent to PHP and here are the entry point of modify/add/sub:
https://github.com/php/php-src/blob/a555cc0b3d4f745e6d0bb8c595de400a0c728827/ext/date/php_date.c#L103

But be warned, in my memories, the overflow were done by an other external library PHP relies on for date manipulations.

**Comment by Noushid** [comment]
I have same Issue please go through below code.
Code
```
('current ' . Carbon::now()->format('Y-m-d'));//2021-04-30
('sub 1 '.Carbon::now()->subMonths(1)->format('Y-m-d'));
('sub 2 '.Carbon::now()->subMonths(2)->format('Y-m-d'));
('sub 3 '.Carbon::now()->subMonths(3)->format('Y-m-d'));
```

Result
```
current 2021-04-30' 
sub 1 2021-03-30' 
sub 2 2021-03-02' 
sub 3 2021-01-30' 
```

**Comment by lucasmichot** [comment]
@Noushid use `subMonthsNoOverflow`

**Comment by itofficeeu** [comment]
Here another problem in the same league:

$database_format->year('2023')
                                                    ->month('01')
                                                    ->day('31')
                                                    ->hour(12)
                                                    ->minute(0)
                                                    ->second(0)
                                                    ->format('Y-m-d');

returns correctly 2023-01-31

BUT:

$database_format->year('2023')
                                                    ->month('02')
                                                    ->day('31')
                                                    ->hour(12)
                                                    ->minute(0)
                                                    ->second(0)
                                                    ->format('Y-m-d');

returns 2023-03-31. - But I need the corresponding day of the last month, or as this obviously isn't possible in this case, the last day of the month.

I personally see it as an error not providing the last day of the month. But feel free to have another opinion. I will read and listen, and maybe learn.

If not possible, it could be helpful (during development) to get an error in some way.

Maybe there already is a special command to surpass this, which I am tempted to call an anti-pattern?... Any hint on that, now we have a thread here about it?... Just for the next many developer falling into this. Could save them many hours.

PHP is existing since 1995.

Laravel since 2011.

Date formats and manipulating them is a basic thing you need again and again. - Shouldn't that be a good reason to improve this part ?... PHP and Laravel is improved in many ways each year, thanks, but why not improve this?... Please?..

Thank you for constantly improving Laravel, and of course having created it the first place.

**Comment by kylekatarnls** [comment]
`$carbon_date->month($month_number)` is dead code here, as you set to the same month it already is.

If your use case is that simple, you just need:

```php
$carbon_date->day(min($carbon_date->daysInMonth, $day_to_set));
```

(Side note, if the use case is to set year/month/day together, use `setDate()` instead.)

And that is not worth a fix, because you can have what you want easily and because not everyone want to get `$carbon_date->day(min($carbon_date->daysInMonth, $day_to_set))` when they ask for `$carbon_date->day($day_to_set)`.

You're biased by your current need. As said previously, there is no correct success output to the question "Get me the date for the 29th day of February" receiving 2023-02-28 is not a more "finished solution", and 2023-03-01 is not more an "error", you can be the 999th person going complain about overflow behavior with PHP team, trust me, they are well aware of this behavior, and PHP is far from being the only language to handle it this way. No major discovery here.

The more correct output you could get to this question is an error/exception such as `InvalidDateException`, we won't do that in a minor release of backward compatibility reason, but it's an option I consider for a while for the version 3.0.

If what you propose not overflow (get 2023-02-28) that's a "no" for Carbon, we won't replace an incorrect output with an other incorrect output.

I guess for the same reason, PHP team prefer to keep current behavior for backward-compatibility reason since the alternative will also cause trouble in many sitations.

**Comment by king724** [comment]
It is January 31st 2024 and I just got bit by this.

I swapped out my `subMonths(2)` for `subMonthsNoOverflow(2)` to get the proper dates in my tests.

**Comment by kylekatarnls** [comment]
It's a general pitfall working with dates (not Carbon-specific, not even PHP-specific) which is to not anticipate the things that can happen differently when they happen:
- on a 31st
- adding/subtracting month and February is in the way
- before/after a daylight saving time change
- just before/just after midnight

In various business doing various stuffs combining multiple date-time operations, the field of the possible expected behaviors is infinite and there is no one single rule that could work for every business and every cases.

If the test is not mocked to simulate those particular cases, you're likely to discover your business expectations are not met late and the hard way.

I'm crafting some improvements on [Carbonite test tooling](https://github.com/kylekatarnls/carbonite) that will make it easier and test what the output of your tests would look like at some particular dates, provide some random dates and also some dates as the listed above known to trigger edge-cases.

In the meantime, what I could recommend is to mock time as much as you can with whatever you suspect could trigger an edge-case for you:

With Carbon, it looks like:
```php
class FooTest extends TestCase
{
  public function testThing()
  {
    Carbon::setTestNow('2024-01-31 23:12');
    self::assertSame(11, Carbon::now()->subMonthsNoOverflow(2)->month);
  }
}
```
With [Carbonite](https://github.com/kylekatarnls/carbonite)
```php
class FooTest extends TestCase
{
  use BespinTimeMocking;

  #[Freeze('2024-01-31 23:12')]
  public function testThing()
  {
    self::assertSame(11, Carbon::now()->subMonthsNoOverflow(2)->month);
  }
}
```

**Comment by TimtheStew** [comment]
As someone who just got bit by this - the problem is really our gosh darned hubris in attempting to delineate the constant barrage of moments composing our existence that is the problem.

Throw out your clocks and live, free of the burden of the boundaries we put on our time.

Or yeah, use `subMonthsNoOverflow()`
